src/System.Management.Automation/FormatAndOutput/common/BaseOutputtingCommand.cs

```diff
@@ -1006,11 +1006,11 @@ internal override void ProcessPayload(FormatEntryData fed)
 
                 // need to make sure we have matching counts: the header count will have to prevail
                 string[] values = new string[headerColumns];
-                int[] alignment = new int[headerColumns];
+                Span<int> alignment = stackalloc int[headerColumns];
```

Comment:
Same here, `alignment` is passed to another method `GenerateRow`, so it's not safe to use `stackalloc` here.
Plus, `GenerateRow` takes array argument, so another conversion will happen.

Reply:
I should add that Span<T> is _managed_ code.
Before C# 7.2 we have some protections for Span<T> only in runtime.
Since C# 7.2 Roslyn does additional checks at compile time and exclude unapproved usage..

Reply:
I understand the C# compiler can catch an issue like that, but the person who is facing that compilation error might not be the one who used `stackalloc` in the first place.

Imagine I will change the implementation of `GenerateRow` at some later time and my change would actually cause the unsafe reference to happen, and that means I will see the compilation error but have no idea why it fails. After finally figuring out why this could happen, I will need to search all callers of this methods and maybe the callers of those callers and so on to find "Ah, there is a stackalloc here".

Therefore, I think we should have a guideline about when to use `stackalloc`, so things can be less complex.

Reply:
I agree that the error might be confusing to another contributor.  

If we have guidance on when `span` and `stackalloc` can be used it should be clear and easy to follow.

Reply:
>who is facing that compilation error might not be the one who used stackalloc in the first place.

I still have small experience with Span and again don't understand. The compilation error is not related to stackalloc - it is related to Span. My understanding is that Span is managed code and full safe.
The pattern is widely used in CoreFX: 
```c#
Span<int> arr;
if ( len < threashold )
{
    arr = stackalloc int[len]
}
else
{
    arr = new int[len]
}
```
We can safely use the arr in the method, pass it to sub-methods, get a slice of the arr from sub-methods. We'll get errors only if the arr will not be used properly - in an iterator, assign to a heap entity and so on. So if we see Span in parameter definition we explicitly see the limitations.

Reply:
[Guidance](https://github.com/dotnet/corefx/blob/e50ddc30deca8fa7434fcc32991354d5a38b5efd/Documentation/api-guidelines/System.Memory.md#guidance)

Reply:
> The pattern is widely used in CoreFX

@iSazonov Thanks for pointing out it. I quickly searched in corefx code base and did uses like that. It looks I'm being overcautious here :)

`Span<T>` cannot be boxed, so yes, it's more accurate to say the limitation is on `Span<T>`, not directly stemming from `stackalloc`.

The scenario that I can think of where the compilation error actually stems from a `stackalloc` statement would be the pesudo code as follows
```
Span<int> MethodOne(...) {
    Span<int> arr = stackalloc int[len];
    return MethodTwo(arr); // <-- return the span returned from another method
}

Span<int> MethodTwo(Span<int> span, ...) {
    ....
    return span.Slice(...); // <-- return a slice of the passed-in span
}
```
In this example, when `stackalloc` was initially applied, it might work fine because `MethodTwo` didn't directly return a slice of the passed-in span. But if a later change to `MethodTwo` makes it `return span.Slice(...)`, then a compilation error will be thrown, in `MethodOne` I guess (didn't really try myself).

What should be the right use in this example? 
- Not returning a slice of a passed-in span argument?
- Or not directly returning a span result returned from another method when you pass in stack memory as argument? 

We are all new to the new language feature, and this is why I would like some guidance around using `stackalloc` and `Span<>`. I think this discussion is very useful for us all to learn more about this new feature and get a better understanding.

Reply:
Related to Span<T>, it looks like it's not currently supported on Alpine Linux so our build script for Alpine is currently broken...

Reply:
> What should be the right use in this example?

We should be thinking about Span<T> as _managed_ code - it is really managed code. Roslyn (and runtime) strongly track all references on Span and block us if we'll try to return (directly or not directly) Span (slice of the span) to above scope from one where it was defined with stackalloc.
So in your example `return MethodTwo(arr); ` will throw an error in any case.

Reply:
@SteveL-MSFT They don't support Alpine in [2.0](https://blogs.msdn.microsoft.com/dotnet/2017/08/14/announcing-net-core-2-0/) and support in 2.1 [Prview1](https://blogs.msdn.microsoft.com/dotnet/2018/02/27/announcing-net-core-2-1-preview-1/) and perhaps in [Preview2](https://blogs.msdn.microsoft.com/dotnet/2018/04/11/announcing-net-core-2-1-preview-2/)

Reply:
> So in your example return MethodTwo(arr); will throw an error in any case (Span and Slice).

I think you didn't get my point.

What I try to point out is **the use of `stackalloc` might make future changes raise compilation error at an unexpected location.**

Use the same example, say initially methods `MethodOne` and `MethodTwo` looked like this:
```c#
Span<int> MethodOne(...) {
    int[] arr = new int[len];
    return MethodTwo(arr); // <-- return the span returned from another method
}

Span<int> MethodTwo(Span<int> span, ...) {
    int[] array = new int[length];
    ...
    return array.slice(...); // <-- return a slice of an array
}
```
So when you changed to use `stackalloc` in `MethodOne`, it would compiles fine, and the methods became:
```c#
Span<int> MethodOne(...) {
    Span<int> arr = stackalloc int[len];
    return MethodTwo(arr); // <-- return the span returned from another method
}

Span<int> MethodTwo(Span<int> span, ...) {
    int[] array = new int[length];
    ...
    return array.slice(...); // <-- return a slice of an array
}
```
Then later, another contributor changes `MethodTwo` to return a sclice of the pass-in argument `span` like this:
```c#
Span<int> MethodTwo(Span<int> span, ...) {
    ....
    return span.Slice(...); // <-- return a slice of the passed-in span
}
```
And then the contributor **will see the unexpected compilation error from `MethodOne`.**

This is the situation that I want to avoid with some coding guidelines. So what should the guideline be in this situation? Should it be:
- Do not return a slice of a passed-in span argument in a method? (_attempt to prevent this situation from `MehtodTwo`_)
- Or, do not directly return a span result returned from another method when you pass in stack memory as argument to that method? (_attempt to prevent this situation from `MethodOne`_)

Hope I made myself clear this time :)

Reply:
I see your point. I believe we could compare this with changing in a base class - we can catch tons compilation errors in derived classes - this is the same situation. Another example is a changing in a helper class - we can get compilation errors in each call point. 
And it seems like your example is too incredible if we try to imagine a real situation - such a change in the code completely changes its contract as if `Array. Sort` would start to do a `clone-then-sort`. By the way, in the latter case we may get an unpredictable run-time error or nothing worse. So Span gives even more reliable code.

Reply:
I think this is something that we will have to get used to. It is a new and very welcome addition to the .net runtime, and we as developers need to learn to troubleshoot the issues that arises. 
I say let the compilation error happen, and add answers to StackOverflow on how to resolve and understand them.

Reply:
If we use a threshold to select `stackalloc` vs `new` we could use one of two patterns
https://github.com/dotnet/corefx/blob/58af9adc8c6c70df4b7fd158c984508d96baca59/src/Common/src/System/Security/Cryptography/ECDiffieHellmanOpenSsl.Derive.cs#L165
https://github.com/dotnet/corefx/blob/8b70122cef4a165def5224001b38f10967619669/src/Common/src/CoreLib/System/Globalization/CompareInfo.Windows.cs#L141

Also I found interesting sample https://github.com/dotnet/corefx/blob/fcfb556ebf99ed551c02d0146c805e792b264cf2/src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemName.cs#L197 - using `Span<int> temp = stackalloc int[0];`  disable expose outside of the method.